executor/pipeline/reduce.rs

```diff
@@ -0,0 +1,205 @@
+            program.reduction_inputs.iter().map(|reducer| ReducerImpl::build(reducer)).collect();
+        let reused_group = Vec::with_capacity(program.input_group_positions.len());
+        Self {
+            input_group_positions: program.input_group_positions,
+            grouped_aggregates: HashMap::new(),
+            reused_group,
+            sample_reducers,
+        }
+    }
+
+    fn accept<Snapshot: ReadableSnapshot>(
+        &mut self,
+        row: &MaybeOwnedRow<'_>,
+        context: &ExecutionContext<Snapshot>,
+    ) -> Result<(), PipelineExecutionError> {
+        self.reused_group.clear();
+        for pos in &self.input_group_positions {
+            self.reused_group.push(row.get(pos.clone()).to_owned());
+        }
+        if !self.grouped_aggregates.contains_key(&self.reused_group) {
+            self.grouped_aggregates.insert(self.reused_group.clone(), self.sample_reducers.clone());
+        }
+        let reducers = self.grouped_aggregates.get_mut(&self.reused_group).unwrap();
+        for reducer in reducers {
+            reducer.accept(row, context);
+        }
+        Ok(())
+    }
+
+    fn finalise(self) -> Batch {
+        let Self { input_group_positions, sample_reducers, grouped_aggregates, .. } = self;
+        let mut batch =
+            Batch::new((input_group_positions.len() + sample_reducers.len()) as u32, grouped_aggregates.len());
+        let mut reused_row = Vec::with_capacity(input_group_positions.len() + sample_reducers.len());
+        let reused_multiplicity = 1;
+        for (group, reducers) in grouped_aggregates.into_iter() {
+            reused_row.clear();
+            for value in group.into_iter() {
+                reused_row.push(value);
+            }
+            for reducer in reducers.into_iter() {
+                reused_row.push(reducer.finalise());
+            }
+            batch.append(MaybeOwnedRow::new_borrowed(reused_row.as_slice(), &reused_multiplicity))
+        }
+        batch
+    }
+}
+
+trait ReducerAPI {
+    fn new(target: VariablePosition) -> Self;
+
+    fn accept<Snapshot: ReadableSnapshot>(&mut self, row: &MaybeOwnedRow<'_>, context: &ExecutionContext<Snapshot>);
+
+    fn finalise(self) -> VariableValue<'static>;
+}
+
+fn extract_value<Snapshot: ReadableSnapshot>(
+    row: &MaybeOwnedRow<'_>,
+    position: &VariablePosition,
+    context: &ExecutionContext<Snapshot>,
+) -> Option<Value<'static>> {
+    match row.get(position.clone()) {
+        VariableValue::Empty => None,
+        VariableValue::Value(value) => Some(value.clone().into_owned()),
+        VariableValue::Thing(Thing::Attribute(attribute)) => {
+            // As long as these are trivial, it's safe to unwrap
+            let snapshot: &Snapshot = &context.snapshot;
+            let value = attribute.get_value(snapshot, &context.thing_manager).unwrap();
+            Some(value.clone().into_owned())
+        }
+        _ => unreachable!(),
+    }
+}
```

Comment:
Does this have to be an owned value you return? This will induce extra Clones you don't really need I think?

Reply:
One is owned, the other isn't. I can wrap it in `MaybeOwns`. I can also likely make it return the primitive.

Reply:
Should be possible to make this not owned in any branch